cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/upgrade/DataMigration.java

```diff
@@ -0,0 +1,329 @@
+      "-", ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, Constants.SYSTEM_NAMESPACE),
+
+      "gateway", ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, Constants.SYSTEM_NAMESPACE,
+                                 Constants.Metrics.Tag.COMPONENT, Constants.Gateway.METRICS_CONTEXT,
+                                 Constants.Metrics.Tag.HANDLER, Constants.Gateway.STREAM_HANDLER_NAME)
+    );
+  private final String entityTableName;
+  private final String metricsTableName;
+
+  public DataMigration(final CConfiguration cConf, final DatasetFramework dsFramework,
+                       DefaultMetricDatasetFactory factory) {
+
+    LOG.info("Initializing Data Migration.");
+    this.dsFramework = dsFramework;
+    this.entityTableName = cConf.get(MetricsConstants.ConfigKeys.ENTITY_TABLE_NAME,
+                                     UpgradeMetricsConstants.DEFAULT_ENTITY_TABLE_NAME);
+    this.metricsTableName = cConf.get(MetricsConstants.ConfigKeys.METRICS_TABLE_PREFIX,
+                                      UpgradeMetricsConstants.DEFAULT_METRICS_TABLE_PREFIX) + ".agg";
+    aggMetricStore = new DefaultMetricStore(factory, new int[]{Integer.MAX_VALUE});
+  }
+
+  public void migrateMetricsTable(Version cdapVersion) {
+    if (cdapVersion.getVersion().equals("2.6")) {
+      migrateMetricsTableFromVersion26(cdapVersion);
+    } else if (cdapVersion.getVersion().equals("2.7")) {
+      migrateMetricsTableFromVersion27(cdapVersion);
+    } else {
+      LOG.info("Can Migrate only from CDAP versions 2.6 and 2.7 to CDAP-version 2.8");
+    }
+  }
+
+  private void migrateMetricsTableFromVersion26(Version version) {
+    for (String scope : scopes) {
+      EntityTable entityTable = new EntityTable(getOrCreateMetricsTable(String.format("%s.%s", scope, entityTableName),
+                                                                        DatasetProperties.EMPTY));
+      MetricsTable metricsTable = getOrCreateMetricsTable(String.format("%s.%s", scope, metricsTableName),
+                                                          DatasetProperties.EMPTY);
+      migrateMetricsData(entityTable, metricsTable, scope, version);
+    }
+  }
+
+  private void migrateMetricsTableFromVersion27(Version version) {
+    EntityTable entityTable = new EntityTable(getOrCreateMetricsTable(
+      UpgradeMetricsConstants.DEFAULT_ENTITY_TABLE_NAME, DatasetProperties.EMPTY));
+    MetricsTable metricsTable = getOrCreateMetricsTable(metricsTableName, DatasetProperties.EMPTY);
+    migrateMetricsData(entityTable, metricsTable, null, version);
+  }
+
+  private void migrateMetricsData(EntityTable entityTable, MetricsTable metricsTable, String scope, Version version) {
+    MetricsEntityCodec codec = getEntityCodec(entityTable);
+    int idSize = getIdSize(version);
+    Row row;
+    try {
+      Scanner scanner = metricsTable.scan(null, null, null, null);
+      while ((row = scanner.next()) != null) {
```

Comment:
can we log every say 10K records that "migrated XXX records", so that if it is long running - you can see the progress